private/library.rkt

```diff
@@ -0,0 +1,296 @@
+     (if (set-member? collects-dirs b)
+         (let* ([cache (library-setup-cache lib)]
+                [a-pkg (path->pkg a #:cache cache)]
+                [b-pkg (path->pkg b #:cache cache)])
+           (cond
+             [(equal? a-pkg b-pkg)
+              (string<? (directory->pretty-string lib a)
+                        (directory->pretty-string lib b))]
+             [(and a-pkg b-pkg)
+              (string<? a-pkg b-pkg)]
+             [a-pkg
+              #t]
+             [else
+              #f]))
+         #t)]
+    [(set-member? collects-dirs b)
+     #f]
+    [else
+     (path<? a b)]))
+
+(define (directory->pretty-string lib dir #:file [file #f])
+  (define full
+    (if file
+        (build-path dir file)
+        dir))
+  (hash-ref!
+   (library-pretty-cache lib)
+   full
+   (λ ()
+     (string->immutable-string
+      (if (hash-has-key? (library-data-table (library-lib lib)) dir)
+          (path->string full)
+          (path->relative-string/library full #:cache (library-setup-cache lib)))))))
+
+(define (path->normalized-lib-module-path lib pth)
+  (hash-ref!
+   (library-mp-cache lib)
+   pth
+   (λ ()
+     (define rslt
+       (path->module-path pth #:cache (library-setup-cache lib)))
+     (and (normalized-lib-module-path? lib)
+          `(lib ,(string->immutable-string (cadr rslt)))))))
+
+(define (path->writable-module-path lib pth)
+  (or (path->normalized-lib-module-path lib pth)
+      `(file ,(string->immutable-string (path->string pth)))))
+
+(define (directories lib #:sorted? [sorted? #f])
+  (if sorted?
+      (sort (directories lib)
+            (λ (a b)
+              (directory<? lib a b)))
+      (append (hash-keys (library-data-table (library-lib lib)))
+              (set->list (library-collects-dirs lib)))))
+
+(define (directory->enabled+file lib dir)
+  (define data (library-data lib))
+  (define enabled?
+    (cond
+      [(hash-ref (library-data-table data) dir #f)
+       => (λ (excludes)
+            (λ (name)
+              (not (set-member? excludes name))))]
+      [else
+       (define excludes (library-data-collection-based-exclusions data))
+       (λ (name)
+         (define pth (build-path dir name))
+         (not (set-member? excludes (path->normalized-lib-module-path lib pth))))]))
+  (for/list ([name (in-list (if (directory-exists? dir)
+                                (directory-list dir #:build? #f)
+                                '()))]
+             #:when (and (script-file? name)
+                         (file-exists? (build-path dir name))))
+    (cons (enabled? name) name)))
+
+(define (all-enabled-scripts lib)
+  (for*/list ([dir (in-list (directories lib))]
+              [enabled+file (in-list (directory->enabled+file lib dir))]
+              #:when (car enabled+file))
+    (build-path dir (cdr enabled+file))))
+
+(define (removable-directory? lib dir)
+  (and (hash-has-key? (library-lib (library-data-table lib)) dir)
+       (not (equal? user-script-dir dir))))
```

Comment:
Some parts of Quickscript assume `user-script-dir` is in the library. I've preserved the existing behavior of disabling the "remove" button for it, and that also seemed to make sense for package-based scripts.

I thought a little about trying to better ensure `user-script-dir` is always present, perhaps by not putting it in the table as a path but storing its exclusions separately. But what should a user had added some additional directory `A`, then moved their installation or set `PLTQUICKSCRIPTDIR` so that `A` became `user-script-dir`?

Reply:
I think I would remove `PLTQUICKSCRIPTDIR` and replace it with `PLTQUICKSCRIPTLIBRARY` which contains all the necessary information.

Random ideas:
* For non-collection paths, exclusions could also be stored in the scripts directory in an `info.rkt` file, as for collections. This would allow to have relative paths.
* Stored collection-based script paths could *not* be absolute, but prefixed with the collection name, as in `require` paths. That can be quite annoying to handle though.

Reply:
It seems reasonable to specify scripts using the same notation as `require` and then they can be passed to `dynamic-require` to actually load the scripts.

Reply:
I am confused about which parts of the PR the replies in this sub-thread are commenting on.

The code fragment here (defining `removable-directory?`) is used only in the library GUI. It generalizes the existing behavior of disabling the "remove" button for `user-script-directory`.

The question I was trying to ask was whether it would be worth-while to change the `library` representation to make it impossible to represent a state without `user-script-dir` in the library.

> I think I would remove `PLTQUICKSCRIPTDIR` and replace it with `PLTQUICKSCRIPTLIBRARY` which contains all the necessary information.

I'm not sure what this would mean. I think it would be quite inconvenient to set an environment variable with the complete serialized `library-data` structure, especially since the serialized format isn't designed to be written by hand. If I understand correctly that `PLTQUICKSCRIPTDIR` is only intended for development and testing, I would remove it and rely on `PLTUSERHOME` (and `XDG_CONFIG_HOME` on Unix).

> It seems reasonable to specify scripts using the same notation as `require` and then they can be passed to `dynamic-require` to actually load the scripts.

This PR represents a collection-based script as a [`normalized-lib-module-path?`](https://docs.racket-lang.org/raco/collection-search.html#%28def._%28%28lib._setup%2Fcollection-search..rkt%29._normalized-lib-module-path~3f%29%29). (Directories containing collection-based scripts are represented as complete paths, but these are found via `find-relevant-directories` and never stored by this library.)

Reply:
> I am confused about which parts of the PR the replies in this sub-thread are commenting on.
> 
> The code fragment here (defining `removable-directory?`) is used only in the library GUI. It generalizes the existing behavior of disabling the "remove" button for `user-script-directory`.
> 
> The question I was trying to ask was whether it would be worth-while to change the `library` representation to make it impossible to represent a state without `user-script-dir` in the library.

I'm not sure that's a good idea, but you're welcome to try something :)

> > I think I would remove `PLTQUICKSCRIPTDIR` and replace it with `PLTQUICKSCRIPTLIBRARY` which contains all the necessary information.
> 
> I'm not sure what this would mean. I think it would be quite inconvenient to set an environment variable with the complete serialized `library-data` structure, especially since the serialized format isn't designed to be written by hand. If I understand correctly that `PLTQUICKSCRIPTDIR` is only intended for development and testing, I would remove it and rely on `PLTUSERHOME` (and `XDG_CONFIG_HOME` on Unix).

My bad, I meant `PLTQUICKSCRIPTLIBRARYPATH`. But yes, I don't think anyone but me is actually using this variable. Yes, I think we can rely only on PLTUSERHOME.

> 
> > It seems reasonable to specify scripts using the same notation as `require` and then they can be passed to `dynamic-require` to actually load the scripts.
> 
> This PR represents a collection-based script as a [`normalized-lib-module-path?`](https://docs.racket-lang.org/raco/collection-search.html#%28def._%28%28lib._setup%2Fcollection-search..rkt%29._normalized-lib-module-path~3f%29%29). (Directories containing collection-based scripts are represented as complete paths, but these are found via `find-relevant-directories` and never stored by this library.)

Seems good to me!

Reply:
> > The question I was trying to ask was whether it would be worth-while to change the `library` representation to make it impossible to represent a state without `user-script-dir` in the library.
> 
> I'm not sure that's a good idea, but you're welcome to try something :)

Let's imagine Alice and Bob both have DrRacket on their respective laptops. They both have the default `user-script-dir`, i.e. `~/.config/racket/quickscript/user-scripts/`, and have both added an exclusion for `foo.rkt` in that directory.

 1. Alice uses the GUI to add an additional directory, `/tmp/racket/quickscript/user-scripts/`, which contains two files, `foo.rkt` and `bar.rkt`. Alice also uses the GUI to add an exclusion for `/tmp/racket/quickscript/user-scripts/bar.rkt`. Finally, Alice does `cp ~/.config/racket/racket-prefs.rktd /tmp/racket`.

 2. Instead of the above, Bob does `mv ~/.config/racket /tmp`.

Now Alice and Bob close DrRacket, set `XDG_CONFIG_HOME=/tmp`, and restart DrRacket. Now `user-script-dir` would be `/tmp/racket/quickscript/user-scripts/`.

With the status quo, Alice's DrRacket will have exactly the same script setup as before. Bob's DrRacket, however, will not find any scripts, because it will look in the only complete path it has saved, `/home/bob/.config/racket/quickscript/user-scripts/`, which no longer exists. Furthermore, Bob's library won't have an entry for `user-script-dir`, which several parts of Quickscript seem to assume it will.

If instead we change to Quickscript to treat `user-script-dir` specially, Bob's attempt to move `XDG_CONFIG_HOME` will work as expected. Furthermore, we can treat guarantee that `user-script-dir` is always part of the library. On the other hand, Alice's situation becomes a bit murky. Naively, she might end up with duplicate entries for `/tmp/racket/quickscript/user-scripts/`: we could at least avoid the duplication, and it might make sense to merge the exclusions (so she would end up with exclusions for both `foo.rkt` and `bar.rkt`).

Neither scenario seems especially likely to be common, but I can more easily imagine Bob's scenareo being useful than Alice's, which, combined with guaranteeing that `user-script-dir` will always be present, makes me slightly inclined toward the change.

Reply:
Thanks for the detailed example.

I don't think Alice should not expect things to work well, so I would not consider this scenario legitimate and in need of support.

In Bob's scenario, if I'm Bob  I would *hope* for things to go well, but I wouldn't *expect* it to work well unless it's written explicitly in the docs. 

However, if you think supporting Bob's scenario automatically is desirable, no problem.

You're also correct that `user-script-dir` is currently treated half-specially. I should have avoided creating the directory automatically, to just create it if it doesn't exist when a new script is created. 

If exclusions for non-collection script directories are written within the script directory itself, then moving such a directory would move the exclusions at the same time. There would be no need for handling absolute vs relative paths for exclusions. Bob would then only have to register the new path, and unregister the old one — if `user-script-dir` is stored as absolute path.